### DISCUSSION TITLE
Labs Feedback: @lit-labs/virtualizer

### DISCUSSION THREAD

**Discussion by justinfagnani** [lit/lit#3362] [Labs Feedback]
The [`@lit-labs/virtualizer`](https://www.npmjs.com/package/@lit-labs/virtualizer) package provides viewport-based virtualization (including virtual scrolling) for Lit.

A virtualizer is an element that renders its own children, applying the provided `renderItem` template function to each item in the provided `items` array.

Instead of immediately rendering a child element for every item in the array, a virtualizer renders only enough elements to fill the viewport. As the viewport scrolls or resizes, the virtualizer automatically removes elements that are no longer visible and adds elements that have come into view.

We want to collect your feedback on your experiences with `@lit-labs/virtualizer` as we iterate on it according to our Lit Labs process and move the package towards possible graduation or retirement.

- Status: **Under development**
- Major known issues:
  - _TODO_
- Issues: [[labs/virtualizer]](https://github.com/lit/lit/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+%5Blabs%2Fvirtualizer%5D)
- Docs: [README](https://github.com/lit/lit/tree/main/packages/labs/virtualizer#readme)
- npm: [@lit-labs/virtualizer](https://www.npmjs.com/package/@lit-labs/virtualizer)

Please try `@lit-labs/virtualizer` and tell us about your experience. Leave comments with your thoughts; highlight what issues or missing features you're blocked on; give a link to a project that uses it; etc.

And so we can have some numbers to look at, you can add reactions to this comment to indicate your experience with this package:

🚀 I have tried this package and it works great. I need it. Ship it!
👍 I have tried this package and it seems like a good idea.
😕 I have tried this package and would use it, but it has serious bugs or missing features.
👀 I am interested, but haven't tried the package yet.
👎 I'm not interested in this package.

Note: other reactions like 😄, 🎉, and ❤️ don't mean anything as part of this scheme. You probably mean, and please use, 🚀, 👍, or  👀.

**Comment by badlogic** [comment]
I'm evaluating using this for my BlueSky client, in which I have infinitely scrolling lists of social media posts. There are a few issues that can not be resolved with the current API:

* Prepending items: this is required for bi-directional infinite scrolling. Ie. if the user scrolls to the bottom of a posts list, we load older items and append them to the list. This works fine. If the user scrolls to the top of the posts list, we load newer items and prepend them to the virtualized items list. Ideally, the scroll position relative to what is visible on screen would be retained as is the case with [scroll anchoring](https://developer.mozilla.org/en-US/docs/Web/CSS/overflow-anchor/Guide_to_scroll_anchoring). However, currently, the `scrollTop` of the scroll container stays the same, effectively pushing down the visible items outside the viewport. Trying to manually adjust the `scrollTop` of the scroll containter after awaiting `LitVirtualizer.layoutComplete` does not work sadly, as the layout keeps reflowing after that. For social media feeds, one can make due with having a "jump to top" button to jump to the newest item instead of allowing to scroll up. For a chat application, where the newest items are at the bottom and a user wants to scroll upwards through old messages, scroll anchoring is a must have.
* ~~Frequent re-rendering of items that are visible: I have not yet figured out why this happens, but my assumption was that any item tha is within "range" (so visible, or slightly outside the viewport) does not have to be rerendered. However, in my testing, I see that some items get re-rendered often, even if they are in range. Some even get connected/disconnected from the DOM or entirely recreated. This is not great, as it will re-trigger CSS animations, kill state, and so on.~~ Figured this one out: in my Lit templates, I had some DOM nodes embedded instead of templates. This would trigger constant re-rendering.
* Dynamic content support: posts may contain videos. Those are first represented via thumbnail `<img>` elements (with a fixed size so they can be properly virtualized without needing constant reflows). Upon clicking the thumbnail, the `<img>` is swapped out with a `<video>` or `<iframe>`, and the video playback is started. When the item is scrolled outside the "range", its DOM representation is disconnected from the DOM, which obviously means videoplayback stops. I suppose I could reparent the `<video>`/`<iframe>` to someplace else in the DOM if that happens, so playback continues (if that works at all), but have otherwise not found a solution.
* `scrollIntoView()` is not enough: a common feature in social media apps is timeline position restauration. E.g. the user scrolls through their home feed, then closes the tab and does something else. Later, they come back and reload the page. Their previous position in the timeline should be restored exactly how it looked when they left. With `scrollIntoView()`, that's currently not possible exactly, as I only have the option of scrolling to "item boundaries" and not pixel positions like with `scrollTop`. Sadly, using `scrollTop` on the scroll container can also not be used, as the virtualizer doesn't quote scroll to the saved `scrollTop`.

Other than that it has been extremely easy to swap out my old virtualization "solution" with lit-virtualizer and I was surprised that it just worked (minus the issues above). Great stuff!

**Reply by augustjk** [reply]
Thanks for the feedback! cc @graynorton

**Comment by cuberoot** [comment]
We are using the virtualizer in a product. Here are a couple of things that we encountered:
- It is difficult to implement a view where the newest items are at the bottom and new items are paged in at the top. It is difficult to make the virtualizer stable while the new items are loaded in, especially if the virtualizer is not scrolled to the top when the new items are loaded.
- We would like to see the API for writing layouts stablized and published. In our case, we can compute the exact size of each item in the virtualizer (and they are not all the same). So, the resize observers of the flow layout were just creating jank in the scrolling. We did end up writing our own layouts with the unofficial API.

Great work though! Thanks!

cc @graynorton

**Comment by eric-g-97477** [comment]
Something I am curious about....

A potential use case for this would be to display a table. It is useful for tables to have fixed headers that are at the top and always visible. 

How does the design of the virtualizer take this into account?

Would the table header be a separate component outside of the virtualized list of table elements? If so, I can imagine it could be difficult to get table columns to line up well. Or, would that not be an issue?

**Reply by tonyalaribe** [reply]
I'm facing this issue at the moment. Rendering a table with virtualize, but being unable to match table headers to the content. Especially with dynamic columns which can be added and removed at runtime.

**Comment by majornista** [comment]
@justinfagnani At present, when a focused item within the virtualizer scrolls out of view, it gets recycled to render another item and focus gets lost to the `document.body`. This can lead to some undesirable side-effects:

https://github.com/user-attachments/assets/88dd7c00-27b8-48a3-bbb3-7c6d0b2075f0

This may be solved by keeping the element that has focus within the virtualizer in the DOM when it scrolls out of view.

**Comment by dcapslock** [comment]
Some issues to note, in both circumstances NOT using the inbuilt scroller.

1. I have a circustance where the virtualizer is out of view so no items to render, but comes into view by another element of the page being collpased. In this case no items render until the page is scrolled. In code I can similaute scroll to workaround.
2. When updating items for the virtualizer (filtering outside of the virtualizer) and the list reduces down below enough to fill screen, the virtualizer will emit scrolls which scrolls the page down progressively until at the top of the virtualizer element.
